### ISSUE TITLE
isla-lib's memory.rs has equal branches in if/else

### ISSUE THREAD

**Issue by Trolldemorted** [rems-project/isla#16]
https://github.com/rems-project/isla/blob/7184bcb0326f32152183ac56bb6f53426b9687f2/isla-lib/src/memory.rs#L295-L299

Is that intentional?

**Comment by Alasdair** [comment]
The suspect the intent was probably that the concrete `Val::Bits(_)` case would allow writing to concrete regions of memory. Eventually we might need this for running large amounts of single-threaded code.

**Comment by Trolldemorted** [comment]
My follow-up question would have been whether that breaks code that uses return addresses which were saved on the stack, I suspect that is correct? :|

**Comment by Alasdair** [comment]
Depends what you mean by 'breaks'. Right now it's set up so that memory reads and writes are highly unconstrained, because we want to handle all the weird concurrent situations, like where say another thread could come along and overwrite your return address on the stack. This is handled by imposing constraints on the symbolic memory reads and writes only after we generate traces. A good example would self-modifying code where you actually want the write to be symbolic even though it's going to the memory that holds your program (which otherwise needs to be concretely specified) due to how the instruction-cache semantics work even in the single-threaded case.

Now if you want to evaluate large amounts of code under the assumption there are no other threads, and there is no self-modifying code, then the semantics we have right now is probably not the one you want. However the code is set up so you can have multiple regions of memory each with a different semantics.

**Comment by Trolldemorted** [comment]
>Right now it's set up so that memory reads and writes are highly unconstrained, because we want to handle all the weird concurrent situations, like where say another thread could come along and overwrite your return address on the stack. This is handled by imposing constraints on the symbolic memory reads and writes only after we generate traces.

Thanks for that detail, that explains a lot of peculiarities in isla-lib! I have been prowling through the executor and smt code, and was confused about your solver's trace, and how you stitch it together to determine possible memory access patterns.

> A good example would self-modifying code where you actually want the write to be symbolic even though it's going to the memory that holds your program (which otherwise needs to be concretely specified) due to how the instruction-cache semantics work even in the single-threaded case.

Does isla-lib truly support self-modifying code? Do you use the monomorphize operation for that? If yes, where in the ASL -> Sail -> Sail IR -> Sail IRX chain is that injected? How are the instruction-cache semantics involved in the matter? I had boldly assumed that writes by a core would always be immediately visible for reads by that core, and the instruction pipeline would abort (and retry with a fresh fetch) any instruction it has partially wrongly executed, if the instruction was overwritten by its predecessor(s).

>Now if you want to evaluate large amounts of code under the assumption there are no other threads, and there is no self-modifying code, then the semantics we have right now is probably not the one you want. However the code is set up so you can have multiple regions of memory each with a different semantics.

Yeah I had a look at memory.rs and thought about the scale of adding a memory section which can take both concrete values and symbols, but I don't feel firm enough in isla-lib to estimate whether that would be feasible, since I don't understand the width of your symbols yet. Angr has [bitvectors of arbitrary size](https://docs.angr.io/core-concepts/solver), which should (?) also exist in isla-lib (because sail can have bitvectors of arbitrary sizes), but your `Sym` type doesn't look like it knows what width it has.

>Depends what you mean by 'breaks'

By breaking I mean 'I can't execute continue execution'. If a return address is pushed on the stack, isla-lib will push a `WriteMem` event onto the trace, and that symbol will eventually end up in pc upon the corresponding return. What happens then? Will the symbol be monomorphized (in this case into the only valid interpretation of the symbol) and the execution can continue without incorrect forks?

I was a bit shocked when I saw symbols in my registers while running deterministic code, e.g. when executing

```
FD 7B BF A9    stp x29, x30, [sp, #-0x10]!
```
while the registers `x29`, `x30` and `sp` have concrete values, `sp` will have a symbolic value afterwards. Do you know why that happens? Is it because `sp` is updated with an internal interim result, which happens to be symbolic (with 1 valid interpretation)?


I am sorry to bother you with that many questions, but isla-lib is (to my best knowledge) the only symbolic execution engine which models ISAs as accurately as possible, and having a symbolic executor which understands all oddities of what is hidden in 8000+ page ISA manuals (including page table walks, exceptions, processor modes and whatnot) is elementary for proper validation of OS kernel binaries. I'd love to transpile sail directly into a more general-purpose symbolic execution library in an imperative programming language while preserving the function names and call graph (being able to step through your emulator in an IDE with a debugger attached while it executes an instruction would be awesome), but until we have a sail parser that outputs something I can understand unambiguously isla-lib is the best I have.

**Comment by Alasdair** [comment]
> Does isla-lib truly support self-modifying code? Do you use the monomorphize operation for that? If yes, where in the ASL -> Sail -> Sail IR -> Sail IRX chain is that injected? How are the instruction-cache semantics involved in the matter? I had boldly assumed that writes by a core would always be immediately visible for reads by that core, and the instruction pipeline would abort (and retry with a fresh fetch) any instruction it has partially wrongly executed, if the instruction was overwritten by its predecessor(s).

I do use the monomorphize operation for that. It's inserted in the Sail source, and used to ensure that the result of every instruction fetch is concrete - In the sail-arm model the function Step_CPU has a line:
```
      __currentInstr       = __monomorphize(__fetchA64());
```
For each instruction fetch litmus test I do need to annotate each location that can be modified with the set of possible values it could contain, so I get a reasonable case split.

> Yeah I had a look at memory.rs and thought about the scale of adding a memory section which can take both concrete values and symbols, but I don't feel firm enough in isla-lib to estimate whether that would be feasible, since I don't understand the width of your symbols yet. Angr has bitvectors of arbitrary size, which should (?) also exist in isla-lib (because sail can have bitvectors of arbitrary sizes), but your Sym type doesn't look like it knows what width it has.

The Sym type is just a reference to a z3 variable, and z3 knows the length. You would do something like `solver.declare_const(BitVec(256))` to create one, and `solver.length(sym)` to get the length. To keep things simple (implementation-wise), I decided to only allow concrete bitvectors up to a certain size (usually 64, but can also be 129 for CHERI capabilities) and push any longer bitvectors into the SMT solver. Such bitvectors can indeed be symbolic but only have a single possible interpretation.

> By breaking I mean 'I can't execute continue execution'. If a return address is pushed on the stack, isla-lib will push a WriteMem event onto the trace, and that symbol will eventually end up in pc upon the corresponding return. What happens then? Will the symbol be monomorphized (in this case into the only valid interpretation of the symbol) and the execution can continue without incorrect forks?

Right now, the PC will probably end up completely unconstrained, for my use concurrency litmus test use case (https://isla-axiomatic.cl.cam.ac.uk) I essentially want to be executing each thread independently under the assumption that I have arbitrarily many other threads which are interfering with absolutely everything I do. In this world if I write a value to memory and immediately read the same location from the same thread, I can get any possible value because I'm assuming this arbitrary interference. Anything less would be risk being unsound because I might rule out behavior that would be allowed by the axiomatic concurrency model.

Our approach internally at the moment for evaluating larger pieces of operating system code using Isla is to evaluate each instruction we see in the kernel binary in a very unconstrained way, which gives us a big set of traces with lots symbolic values, but then we interpret and simplify the traces filling in the memory read and write events with the concrete values we want, before combining these simplified traces to get the behavior of basic blocks in a single-threaded setting.

Adding memory regions that assume no interference would be pretty easy I think, so I could look at doing that.

> FD 7B BF A9    stp x29, x30, [sp, #-0x10]!

If I set SP_EL0 to 0x80000000 and run this I end up with a single trace containing `(write-reg |SP_EL0| nil #x000000007ffffff0)`. The TakeReset function I use to initialize the processor will set a whole bunch of registers to their architecturally specified reset values, which is often unknown. I'd check to make sure nothing like that is happening.

**Comment by Trolldemorted** [comment]
>If I set SP_EL0 to 0x80000000 and run this I end up with a single trace containing (write-reg |SP_EL0| nil #x000000007ffffff0). The TakeReset function I use to initialize the processor will set a whole bunch of registers to their architecturally specified reset values, which is often unknown. I'd check to make sure nothing like that is happening.

Interesting, I guess I have set (at least) one of the registers to the wrong value then. Which registers do I have to set to which values so that `zTakeReset` does not fork and does not fail asserts (right now it does both for me)? I guess you have a sail function that does that for you?

>Adding memory regions that assume no interference would be pretty easy I think, so I could look at doing that.

It's great to hear that you think it is easy, because I was suspecting it is not. I don't understand your symbol lifecycle yet, my naive assumption was it would be roughly like this:
- a symbol gets created by a read from an uninitialized/volatile source
- as long as that symbol is on the executor's current stack or inside a register (or in an rw memory region if you implement that), it does not go "out of scope" (meaning it stays in the trace and z3's world of things)
- derivated symbols may be created by having a computation of the symbol with another symbol, an immediate value, or itself
- the symbol must not go "out of scope" until all derivated symbols (and their derivatives) have gone "out of scope"
- if a symbol goes "out of scope", it has no further burden on the execution, meaning all corresponding memory is freed and it is removed from z3's world of things

Since the derivation graph of symbols is acyclic, putting them into `Arc`s would be my first intuition, but apparently you have managed to model everything with lifetimes. If I understand your symbolic handling correctly, a newly derived symbol becomes a new AST, which `Z3_inc_ref`s the z3 refcount of its parent (and `Z3_dec_ref`s when it is dropped), and is saved in `Solver`'s corresponding HashMap (`decls`, `func_decls`, `enum_map`). It is also pushed onto the trace, which means solvers created from checkpoints will inherit the AST.

How are the ASTs freed after they go out of scope?

**Comment by bacam** [comment]
I should mention the ISA test generator here, which uses the callbacks in the `memory` module to enforce sequential memory semantics during symbolic execution by adding extra constraints for the SMT solver.  In particular, this means that I can inject instructions into memory using a constraint and `monomorphize` will find them.  In hindsight I'm not sure that this is actually much better than skipping instruction fetch entirely and imposing memory constraints afterwards, but it works well enough.

The test generator isn't currently public because it contains a few details about an experimental architecture which is only just being published, but I could make the memory callbacks available if you're interested.

**Comment by Trolldemorted** [comment]
>The test generator isn't currently public because it contains a few details about an experimental architecture which is only just being published, but I could make the memory callbacks available if you're interested.

That definitely sounds interesting! Do you know when the whole thing will be made available to the general public?

**Comment by Trolldemorted** [comment]
Solved along with #21
